compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala

```diff
@@ -60,14 +60,18 @@ object JavaNullInterop {
     if (sym.name == nme.TYPE_ || sym.isAllOf(Flags.JavaEnumValue))
       // Don't nullify the `TYPE` field in every class and Java enum instances
       tp
-    else if (sym.name == nme.toString_ || sym.isConstructor)
+    else if (sym.name == nme.toString_ || sym.isConstructor || hasNotNull(sym))
       // Don't nullify the return type of the `toString` method and constructors
+      // or it has a NotNull annotation
       nullifyParamsOnly(tp)
     else
       // Otherwise, nullify everything
       nullifyType(tp)
   }
 
+  private def hasNotNull(sym: Symbol)(implicit ctx: Context): Boolean =
```

Comment:
Let's rename this to `hasNotNullAnnot`, so that it's clear that it checks for annotations.

Reply:
I will push this in the next commit